pkg/noderesourcetopology/cache/store.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+
+	"sigs.k8s.io/scheduler-plugins/pkg/noderesourcetopology/stringify"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+
+	"github.com/k8stopologyawareschedwg/podfingerprint"
+)
+
+var zeroQty resource.Quantity
+
+func init() {
+	zeroQty = resource.MustParse("0")
+}
+
+// nrtStore maps the NRT data by node name. It is not thread safe and needs to be protected by a lock.
+// data is intentionally copied each time it enters and exists the store. E.g, no pointer sharing.
+type nrtStore map[string]*topologyv1alpha1.NodeResourceTopology
+
+// newNrtStore creates a new nrtStore and initializes it with copies of the provided Node Resource Topology data.
+func newNrtStore(nrts []*topologyv1alpha1.NodeResourceTopology) nrtStore {
+	data := make(map[string]*topologyv1alpha1.NodeResourceTopology, len(nrts))
+	for _, nrt := range nrts {
+		data[nrt.Name] = nrt.DeepCopy()
+	}
+	klog.V(6).InfoS("nrtcache: initialized nrtStore", "objects", len(data))
+	return data
+}
+
+// GetByNode returns a copy of the stored Node Resource Topology data for the given node,
+// or nil if no data is associated to that node.
+func (nrs nrtStore) GetByNode(nodeName string) *topologyv1alpha1.NodeResourceTopology {
+	obj, ok := nrs[nodeName]
+	if !ok {
+		klog.V(3).InfoS("nrtcache: missing cached NodeTopology", "node", nodeName)
+		return nil
+	}
+	return obj.DeepCopy()
+}
+
+// Update adds or replace the Node Resource Topology associated to a node. Always do a copy.
+func (nrs nrtStore) Update(nrt *topologyv1alpha1.NodeResourceTopology) {
```

Comment:
yeah, that's exactly the problem here :\ I guess I'll wrap the map in a struct and use a pointer receiver

Reply:
Or I can just use a pointer receiver. Checking.

Reply:
No obvious solution yet. I see examples of similar types in the scheduler framework (https://github.com/kubernetes/kubernetes/blob/v1.26.0-beta.0/pkg/scheduler/framework/runtime/registry.go#L70) I'll keep an eye and look for options, but both pointer receivers and especially wrapping in structs seems not great alternatives here.

Reply:
yup, it's because map is a pointer underneath, so functionally it can ensure correctness. But for concert type, if we're updating it, using a pointer receiver is a must.

Reply:
Right. I guess then I'll have to wrap the maps in structs, it seems to me the somehow less clunky approach.